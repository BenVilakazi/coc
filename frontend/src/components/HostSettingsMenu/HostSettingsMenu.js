import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import config from '../../config';

const propTypes = {
  settingsComponentList: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string,
      component: PropTypes.func,
    }),
  ).isRequired,
};

const { extraLargeDesktopWidth } = config.breakpoint.hostBreakpoints;

const SettingsMenu = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  width: 250px;
  background-color: var(--primary-background-color);
  display: flex;
  flex-direction: column;

  .host-settings-header {
    display: flex;
    align-items: flex-end;
    height: 50px;
  }

  .host-settings-header h3 {
    font-size: 1.5rem;
    line-height: 0.9rem;
    color: var(--secondary-text-color);
  }

  @media (min-width: 800px) {
    width: 400px;

    .host-settings-header {
      height: 130px;
    }

    .host-settings-header h3 {
      font-size: 3.5rem;
      line-height: 2.2rem;
    }
  }

  @media (min-width: 2000px) {
    width: 600px;

    .host-settings-header h3 {
      font-size: 5rem;
      line-height: 3rem;
    }
  }

  @media (min-width: ${extraLargeDesktopWidth}) {
    width: 1100px;

    .host-settings-header {
      height: 200px;
    }

    .host-settings-header h3 {
      font-size: 8rem;
      line-height: 5rem;
    }
  }
`;

const Header = styled.header`
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
`;

function useSettingsMenuHandler(components) {
  const initialState = components.map(() => 'enabled');
  const [accordionSettings, setAccordionSettings] = useState(initialState);

  const anyAreOpen = accordionSettings.some((setting) => setting === 'open');

  function resetAccordions() {
    setAccordionSettings(initialState);
  }

  function handleAccordionClick(settingIndex) {
    const newSettings = accordionSettings.map((setting, index) =>
      index === settingIndex ? 'open' : 'disabled',
    );
    setAccordionSettings(newSettings);
  }

  function makePopulatedAccordion(Accordion, index) {
    return (
      <Accordion
        key={Accordion.name}
        accordionState={accordionSettings[index]}
        onClickActions={{
          open: resetAccordions,
          enabled: () => handleAccordionClick(index),
          disabled: resetAccordions,
        }}
      />
    );
  }

  function makePopulatedButton(Button) {
    return (
      <Button
        key={Button.name}
        isEnabled={!anyAreOpen}
        onDisabledClick={resetAccordions}
      />
    );
  }

  const populatedComponents = components.map(({ type, component }, index) => {
    switch (type) {
      case 'accordion':
        return makePopulatedAccordion(component, index);

      case 'button':
        return makePopulatedButton(component);

      default:
        throw new Error(
          `Component type: '${type}' not found in HostSettingsMenu component builder`,
        );
    }
  });

  return [populatedComponents, resetAccordions];
}

function HostSettingsMenu({ settingsComponentList }) {
  const [components, resetAccordions] = useSettingsMenuHandler(
    settingsComponentList,
  );

  function onOutsideClick(event) {
    if (event.currentTarget === event.target) {
      resetAccordions();
    }
  }

  return (
    <SettingsMenu onClick={onOutsideClick}>
      <Header className="host-settings-header" onClick={resetAccordions}>
        <h3>SETTINGS</h3>
      </Header>

      {components}
    </SettingsMenu>
  );
}

HostSettingsMenu.propTypes = propTypes;

export default HostSettingsMenu;
