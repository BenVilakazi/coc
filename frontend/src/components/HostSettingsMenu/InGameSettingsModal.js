import React from 'react';

import HostSettingsMenu from './HostSettingsMenu';
import PlayerKicker from './SettingsSubComponents/PlayerKicker';
import JoinCodeHider from './SettingsSubComponents/JoinCodeHider';
import JoinCodeShuffler from './SettingsSubComponents/JoinCodeShuffler';
import SkipPlayerButton from './SettingsSubComponents/SkipPlayerButton';
import ToggleFullscreen from './SettingsSubComponents/ToggleFullscreen';

export default function PreGameSettingsModal() {
  return (
    <HostSettingsMenu
      settingsComponentList={[
        { type: 'button', component: SkipPlayerButton },
        { type: 'button', component: JoinCodeHider },
        { type: 'accordion', component: PlayerKicker },
        { type: 'button', component: JoinCodeShuffler },
        { type: 'button', component: ToggleFullscreen },
      ]}
    />
  );
}
