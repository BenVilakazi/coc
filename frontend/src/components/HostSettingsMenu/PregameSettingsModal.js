import React from 'react';

import HostSettingsMenu from './HostSettingsMenu';
import PlayerKicker from './SettingsSubComponents/PlayerKicker';
import JoinCodeHider from './SettingsSubComponents/JoinCodeHider';
import JoinCodeShuffler from './SettingsSubComponents/JoinCodeShuffler';
import ToggleFullscreen from './SettingsSubComponents/ToggleFullscreen';

export default function PreGameSettingsModal() {
  return (
    <HostSettingsMenu
      settingsComponentList={[
        // if you give a wrong 'type' here you will get a "cannot read property
        // 'undefined' of undefined" error in any accordions that you've typed
        // as 'button' and a bunch of propType errors
        { type: 'button', component: JoinCodeHider },
        { type: 'accordion', component: PlayerKicker },
        { type: 'button', component: JoinCodeShuffler },
        { type: 'button', component: ToggleFullscreen },
      ]}
    />
  );
}
