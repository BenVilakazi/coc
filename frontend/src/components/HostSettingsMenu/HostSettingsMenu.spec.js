import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import HostSettingsMenu from './HostSettingsMenu';
import HostProvider from '../../contexts/HostContext/HostContext';

/* eslint-disable react/prop-types */

describe('HostSettingsMenu', () => {
  describe('rendering', () => {
    it('renders the settings heading', () => {
      render(
        <HostProvider>
          <HostSettingsMenu settingsComponentList={[]} />
        </HostProvider>,
      );

      expect(screen.getByText('SETTINGS')).toBeInTheDocument();
    });

    it('renders settings sub-components of type: accordion', () => {
      function testComponent() {
        return <p>TEST TEXT</p>;
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'accordion', component: testComponent },
            ]}
          />
        </HostProvider>,
      );

      expect(screen.getByText('TEST TEXT')).toBeInTheDocument();
    });

    it('renders settings sub-components of type: button', () => {
      function testComponent() {
        return <p>TEST TEXT</p>;
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'button', component: testComponent },
            ]}
          />
        </HostProvider>,
      );

      expect(screen.getByText('TEST TEXT')).toBeInTheDocument();
    });

    it("throws an error and does not render when a sub-component type isn't expected", () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      function ComponentNotDefinedInTheSwitch() {
        return <p>I SHOULD NOT SHOW UP</p>;
      }

      expect(() => {
        render(
          <HostProvider>
            <HostSettingsMenu
              settingsComponentList={[
                { type: 'foo', component: ComponentNotDefinedInTheSwitch },
              ]}
            />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();

      expect(
        screen.queryByText('I SHOULD NOT SHOW UP'),
      ).not.toBeInTheDocument();
    });

    it('matches the expected snapshot', () => {
      const tree = renderer
        .create(
          <HostProvider>
            <HostSettingsMenu settingsComponentList={[]} />
          </HostProvider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    it('closes a given accordion component when a click occurs outside of that accordion', () => {
      function accordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT' : 'CLOSED TEXT'}
          </button>
        );
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'accordion', component: accordionComponent },
            ]}
          />
        </HostProvider>,
      );

      // open the option list and verify that it is open
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT' }));
      expect(screen.getByText('OPEN TEXT')).toBeInTheDocument();

      // click on the settings header and verify that the option list is now closed
      userEvent.click(screen.getByText('SETTINGS'));
      expect(screen.queryByText('OPEN TEXT')).not.toBeInTheDocument();
    });

    it('disables all settings buttons when an accordion is open', () => {
      function accordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT' : 'CLOSED TEXT'}
          </button>
        );
      }

      const enabledButtonClick = jest.fn();

      function buttonComponent({ isEnabled, onDisabledClick }) {
        return (
          <button
            type="button"
            isEnabled
            onClick={isEnabled ? enabledButtonClick : onDisabledClick}
          >
            TEST BUTTON
          </button>
        );
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'accordion', component: accordionComponent },
              { type: 'button', component: buttonComponent },
            ]}
          />
        </HostProvider>,
      );

      // click the test button once and verify that its function fires
      userEvent.click(screen.getByRole('button', { name: 'TEST BUTTON' }));
      expect(enabledButtonClick).toHaveBeenCalledTimes(1);

      // open the accordion
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT' }));
      expect(screen.getByText('OPEN TEXT')).toBeInTheDocument();

      // click the test button again and verify that its function doesn't fire
      userEvent.click(screen.getByRole('button', { name: 'TEST BUTTON' }));
      expect(enabledButtonClick).toHaveBeenCalledTimes(1);
    });

    it('closes an open accordion when a settings button is clicked', () => {
      function accordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT' : 'CLOSED TEXT'}
          </button>
        );
      }

      const enabledButtonClick = jest.fn();

      function buttonComponent({ isEnabled, onDisabledClick }) {
        return (
          <button
            type="button"
            isEnabled
            onClick={isEnabled ? enabledButtonClick : onDisabledClick}
          >
            TEST BUTTON
          </button>
        );
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'accordion', component: accordionComponent },
              { type: 'button', component: buttonComponent },
            ]}
          />
        </HostProvider>,
      );

      // open the accordion
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT' }));
      expect(screen.getByText('OPEN TEXT')).toBeInTheDocument();

      // click the test button and verify that the accordion is closed
      userEvent.click(screen.getByRole('button', { name: 'TEST BUTTON' }));
      expect(screen.queryByText('OPEN TEXT')).not.toBeInTheDocument();
    });

    it('disables all other settings accordions when an accordion is open', () => {
      function accordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT 1' : 'CLOSED TEXT 1'}
          </button>
        );
      }

      function secondAccordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT 2' : 'CLOSED TEXT 2'}
          </button>
        );
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'accordion', component: accordionComponent },
              { type: 'accordion', component: secondAccordionComponent },
            ]}
          />
        </HostProvider>,
      );

      // open the first accordion
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT 1' }));
      expect(screen.getByText('OPEN TEXT 1')).toBeInTheDocument();

      // click the second accordion and verify that it doesn't open
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT 2' }));
      expect(screen.queryByText('OPEN TEXT 2')).not.toBeInTheDocument();
    });

    it('closes an open accordion when another accordion is clicked', () => {
      function accordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT 1' : 'CLOSED TEXT 1'}
          </button>
        );
      }

      function secondAccordionComponent({ accordionState, onClickActions }) {
        return (
          <button type="button" onClick={onClickActions[accordionState]}>
            {accordionState === 'open' ? 'OPEN TEXT 2' : 'CLOSED TEXT 2'}
          </button>
        );
      }

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'accordion', component: accordionComponent },
              { type: 'accordion', component: secondAccordionComponent },
            ]}
          />
        </HostProvider>,
      );

      // open the first accordion
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT 1' }));
      expect(screen.getByText('OPEN TEXT 1')).toBeInTheDocument();

      // click the second accordion and verify that the first accordion closes
      userEvent.click(screen.getByRole('button', { name: 'CLOSED TEXT 2' }));
      expect(screen.queryByText('OPEN TEXT 1')).not.toBeInTheDocument();
    });
  });

  describe('propTypes', () => {
    it('logs an error when not given the settingsComponentList prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <HostProvider>
            <HostSettingsMenu />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it("logs an error when given a settingsComponentList prop that isn't an array", () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <HostProvider>
            <HostSettingsMenu settingsComponentList={false} />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when given a settingsComponentList prop that contains a non-object in the array', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <HostProvider>
            <HostSettingsMenu settingsComponentList={[false]} />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when given a settingsComponentList prop that contains an object without a type in the array', () => {
      function ComponentNotDefinedInTheSwitch() {
        return <p>I SHOULD NOT SHOW UP</p>;
      }

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <HostProvider>
            <HostSettingsMenu
              settingsComponentList={[
                { component: ComponentNotDefinedInTheSwitch },
              ]}
            />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when given a settingsComponentList prop that contains an object without a component in the array', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <HostProvider>
            <HostSettingsMenu settingsComponentList={[{ type: 'button' }]} />
          </HostProvider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not log an error when given a settingsComponentList prop that contains only objects with type and component in the array', () => {
      function ComponentNotDefinedInTheSwitch() {
        return <p>I SHOULD NOT SHOW UP</p>;
      }

      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <HostSettingsMenu
            settingsComponentList={[
              { type: 'button', component: ComponentNotDefinedInTheSwitch },
            ]}
          />
        </HostProvider>,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });
  });
});
