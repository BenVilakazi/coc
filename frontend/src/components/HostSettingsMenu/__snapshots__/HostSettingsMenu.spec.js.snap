// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostSettingsMenu rendering matches the expected snapshot 1`] = `
.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  width: 250px;
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .host-settings-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-settings-header h3 {
  font-size: 1.5rem;
  line-height: 0.9rem;
  color: var(--secondary-text-color);
}

.c1 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

@media (min-width:800px) {
  .c0 {
    width: 400px;
  }

  .c0 .host-settings-header {
    height: 130px;
  }

  .c0 .host-settings-header h3 {
    font-size: 3.5rem;
    line-height: 2.2rem;
  }
}

@media (min-width:2000px) {
  .c0 {
    width: 600px;
  }

  .c0 .host-settings-header h3 {
    font-size: 5rem;
    line-height: 3rem;
  }
}

@media (min-width:3200px) {
  .c0 {
    width: 1100px;
  }

  .c0 .host-settings-header {
    height: 200px;
  }

  .c0 .host-settings-header h3 {
    font-size: 8rem;
    line-height: 5rem;
  }
}

<div
  className="c0"
  onClick={[Function]}
>
  <header
    className="c1 host-settings-header"
    onClick={[Function]}
  >
    <h3>
      SETTINGS
    </h3>
  </header>
</div>
`;
